File: apps/core/src/audit-log.ts
import fs from "node:fs/promises";
import path from "node:path";
import type { ExecutionAudit } from "@nova/contracts";

export class AuditLog {
  constructor(private readonly filePath: string) {}

  async append(audit: ExecutionAudit): Promise<void> {
    await fs.mkdir(path.dirname(this.filePath), { recursive: true });
    const line = `${JSON.stringify(audit)}\n`;
    await fs.appendFile(this.filePath, line, "utf-8");
  }

  async list(limit = 100): Promise<ExecutionAudit[]> {
    try {
      const raw = await fs.readFile(this.filePath, "utf-8");
      const lines = raw
        .split("\n")
        .map((line) => line.trim())
        .filter(Boolean);
      return lines
        .slice(-limit)
        .map((line) => JSON.parse(line) as ExecutionAudit)
        .reverse();
    } catch {
      return [];
    }
  }
}

File: python/nova_core/audit_log.py
from __future__ import annotations

import json
from pathlib import Path

from .models import ExecutionAudit


class AuditLog:
    def __init__(self, file_path: Path) -> None:
        self.file_path = file_path

    def append(self, audit: ExecutionAudit) -> None:
        self.file_path.parent.mkdir(parents=True, exist_ok=True)
        with self.file_path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(audit.model_dump()) + "\n")

    def list(self, limit: int = 100) -> list[ExecutionAudit]:
        if not self.file_path.exists():
            return []

        lines = [line.strip() for line in self.file_path.read_text("utf-8").splitlines()]
        lines = [line for line in lines if line]
        recent = lines[-limit:]
        recent.reverse()
        return [ExecutionAudit.model_validate_json(line) for line in recent]

